### ISSUE TITLE
v:windowid cannot be modified by UIs

### ISSUE THREAD

**Issue by jalvesaq** [neovim/neovim#3626]
The value of `v:windowid` is not defined. In a terminal emulator, I can use the value of `$WINDOWID` instead, and @equalsraf has just implemented `$WINDOWID` on Neovim-Qt too (see: https://github.com/equalsraf/neovim-qt/issues/64). Perhaps the variable `v:windowid` is superfluous and can be deleted.

This issue is somewhat related with https://github.com/neovim/neovim/issues/2681

**Comment by Pyrohh** [comment]
What if there are multiple external interfaces hooked up to an nvim master process, i.e., multiple windows? cc @equalsraf

Also, do you have a use-case in mind?

**Comment by jalvesaq** [comment]
If there are multiple external interfaces connected to an nvim master process, I think each interface should have its own `$WINDOWID`, but I don't have user cases in mind.

User cases for single Neovim instances are:

   - [Nvim-R]: Send a line of code to R (have to raise R Console window to paste the text) and raise Neovim window again after the text was pasted.

   - [vimtex] and [Nvim-R]: SyncTeX backward jump, that is, do a command in the PDF viewer (double click, CTRL-click, etc, depending on the viewer) to send a message to the editor requesting to move the cursor to the line corresponding to the position in the PDF and to bring its own window to the foreground.

[Nvim-R]: https://github.com/jalvesaq/Nvim-R
[vimtex]: https://github.com/lervag/vimtex

**Comment by jalvesaq** [comment]
Any decision about this? In my opinion, we should not worry about multiple interfaces connected to a single neovim process. I think neovim should simply set either `v:windowid` or `$WINDOWID` in each interface and let the application or plugin to solve the problem of deciding which interface to activate.

**Comment by justinmk** [comment]
> Any decision about this?

@jalvesaq  Isn't this something for UIs to set? For TUI it is meaningless. So what action is needed from the core to close this issue?

**Comment by jalvesaq** [comment]
Yes, I think the issue can be closed here. The problem should be fixed by Neovim-Qt and other interfaces.

**Comment by justinmk** [comment]
Re-opening this per https://github.com/equalsraf/neovim-qt/issues/64#issuecomment-211991880

**Comment by equalsraf** [comment]
Is it ok to just set v:windowid as readwrite (VV_RO_SBX more likely) or should we add some API for this.

**Comment by justinmk** [comment]
> Is it ok to just set v:windowid as readwrite (VV_RO_SBX more likely)

I think that's ok for now. It avoids proliferation of ad-hoc alternatives in the meantime.

**Comment by equalsraf** [comment]
Starting a PR, as a followup it occurs to me that:

1. In Vim if +X11 is set then v:windowid is also set. In our case the TUI could set v:windowid to $WINDOWID
1. Should we reset v:windowid=0 once all UIs are detached?
1. realized there is an API `vim_get_vvar()` but none to set the value, is this intentional?

**Comment by bfredl** [comment]
> In our case the TUI could set v:windowid to $WINDOWID

Soon the tui will communicate exclusively over the API as well. We should have `ui_set_windowid` request (if we try to do something semantic like tracking focus between different frontends) or at least `vim_set_vvar` (there are other mutable ones like v:errors )

**Comment by justinmk** [comment]
> We should have ui_set_windowid

:+1: Though I will repeat my mantra "there is no reason for UI to be a separate concept (thus why not `vim_set_windowid`)".

**Comment by bfredl** [comment]
well it is not more separate concept than "api/buffer" is a separate concept: a vim instance can have more than one ui connected, each which can have a different windowid (not that were going to use that info soonish, but it is good to be forward compatible), which `ui_set_windowid` indicates. The _implementation_ of the ui api is unneccesarily different, that is true, but #4432 (the first commit, I should break that out), will fix that.

**Comment by justinmk** [comment]
> well it is not more separate concept than "api/buffer" is a separate concept

:+1: I was conflating `ui_foo` with the division of "UI events/clients and non-UI clients/clients".

> The implementation of the ui api is unneccesarily different, that is true, but #4432 (the first commit, I should break that out), will fix that.

:+1:

**Comment by justinmk** [comment]
Merged https://github.com/neovim/neovim/pull/4608 to provide a modicum of support for legacy scripts. Keeping this issue open for the valid points mentioned in https://github.com/neovim/neovim/issues/3626#issuecomment-212027679 and https://github.com/neovim/neovim/issues/3626#issuecomment-212076995

**Comment by equalsraf** [comment]
Checking vim_set_vvar(), tried to implement it like this

```C
Object vim_set_vvar(String name, Object value, Error *err)
{
  return dict_set_value(&vimvardict, name, value, false, err);
}
```

but getting an API error `Dictionary is locked`. Which seems to be locked on [startup](https://github.com/neovim/neovim/blob/master/src/nvim/eval.c#L477).

Took a quick look over `set_var()` in eval.c which does not seem to check the global lock like `dict_set_value()` does, but also does [other things](https://github.com/neovim/neovim/blob/master/src/nvim/eval.c#L18076) like enforcing the correct type.for v: entries, but `set_var()` is a static function in eval.c and works for all vars not just v:.
